#Mount Drive

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Load the Database using sqlite

In [2]:
import sqlite3
import pandas as pd

conn = sqlite3.connect('/content/drive/MyDrive/EDA PROJ/ProjectData/gas_monitoring.db')

# Extract the Data into a DataFrame

In [3]:
# Replace 'sensor_data' with the actual table name you found in Step 1
query = "SELECT * FROM gas_monitoring"

# Load the data into a Pandas DataFrame
df = pd.read_sql_query(query, conn)

# Close the connection since we have the data in memory now
conn.close()

# Look at the first 5 rows to see what you're dealing with
display(df.head(50))

,Time of Day,Temperature,Humidity,CO2_InfraredSensor,CO2_ElectroChemicalSensor,MetalOxideSensor_Unit1,MetalOxideSensor_Unit2,MetalOxideSensor_Unit3,MetalOxideSensor_Unit4,CO_GasSensor,Session ID,HVAC Operation Mode,Ambient Light Level,Activity Level
0,morning,19.63,53.83,125.486389,571.089440,478.554958,735.850412,654.625253,565.051969,2.0,7241,off,very_dim,Low Activity
1,morning,292.81,53.69,126.343018,575.789501,491.955151,740.282738,655.734327,557.078486,1.0,7241,ventilation_only,bright,Low Activity
2,morning,19.62,54.25,126.560695,585.543111,505.560808,737.112906,649.962421,558.065196,1.0,7241,eco_mode,None,Low Activity
3,morning,20.10,50.48,113.504877,597.449961,515.142272,744.020651,676.150075,600.222210,NaN,7241,eco_mode,very_bright,High Activity
4,morning,19.89,48.42,92.766225,613.654875,535.664558,770.265441,720.274019,625.731124,1.0,7241,heating_active,moderate,LowActivity
5,afternoon,19.82,47.71,79.489742,616.349705,557.929178,779.316801,738.876591,664.886531,1.0,7241,Ventilation_Only,very_bright,Low Activity
6,night,20.04,48.42,118.293529,582.624643,474.777417,736.890795,694.939447,628.884980,1.0,6431,heating_active,bright,Low Activity
7,night,19.29,52.67,136.292680,574.121555,474.924866,721.380249,680.511012,622.275067,1.0,7963,maintenance_mode,bright,Moderate Activity
8,afternoon,292.59,47.92,74.100322,624.577270,564.745091,798.290403,753.467986,674.278245,0.0,7241,maintenance_mode,very_bright,Low Activity
9,night,19.81,-48.34,8.964022,595.833260,464.828130,737.958044,687.714744,622.836602,NaN,2105,ventilation_only,None,Low Activity


In [4]:
print(f"Dataset contains {df.shape[0]} rows and {df.shape[1]} columns.\n")

print("--- Data Info ---")
df.info()

display(df.describe())

Dataset contains 10000 rows and 14 columns.

--- Data Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Time of Day                10000 non-null  object 
 1   Temperature                10000 non-null  float64
 2   Humidity                   8072 non-null   float64
 3   CO2_InfraredSensor         10000 non-null  float64
 4   CO2_ElectroChemicalSensor  10000 non-null  float64
 5   MetalOxideSensor_Unit1     10000 non-null  float64
 6   MetalOxideSensor_Unit2     8590 non-null   float64
 7   MetalOxideSensor_Unit3     10000 non-null  float64
 8   MetalOxideSensor_Unit4     10000 non-null  float64
 9   CO_GasSensor               9166 non-null   float64
 10  Session ID                 10000 non-null  int64  
 11  HVAC Operation Mode        10000 non-null  object 
 12  Ambient Light Level        8946 non-null

,Temperature,Humidity,CO2_InfraredSensor,CO2_ElectroChemicalSensor,MetalOxideSensor_Unit1,MetalOxideSensor_Unit2,MetalOxideSensor_Unit3,MetalOxideSensor_Unit4,CO_GasSensor,Session ID
count,10000.000000,8072.000000,10000.000000,10000.000000,10000.000000,8590.000000,10000.000000,10000.000000,9166.000000,10000.000000
mean,42.243124,51.643609,109.864352,578.634082,470.932316,728.354408,680.488646,612.247715,1.288239,5427.921400
std,73.974956,21.333894,33.793851,22.573537,51.459025,29.140123,55.953530,43.174354,0.779754,2590.836924
min,6.903485,-49.870000,-17.472184,408.599386,286.825662,611.168916,456.673900,412.791904,0.000000,1374.000000
25%,18.937500,47.660000,96.660377,559.353100,440.035413,709.660122,658.519775,585.511731,1.000000,3074.000000
50%,20.510000,51.390000,112.807103,579.361082,469.146955,726.748925,679.908353,609.526245,1.000000,5214.000000
75%,23.172500,53.840000,124.540226,595.569759,494.770045,744.879789,703.522845,636.012105,2.000000,7395.000000
max,307.070000,198.900000,237.873938,637.895790,632.891693,807.732943,906.213097,736.112361,4.000000,9658.000000


#DATA CLEANING

#Simple data cleaning

In [5]:
print(f"Original rows before dropping duplicates: {len(df)}")
df.drop_duplicates(inplace=True)
print(f"Rows after dropping duplicates: {len(df)}")

display(df.head())

Original rows before dropping duplicates: 10000
Rows after dropping duplicates: 9829


,Time of Day,Temperature,Humidity,CO2_InfraredSensor,CO2_ElectroChemicalSensor,MetalOxideSensor_Unit1,MetalOxideSensor_Unit2,MetalOxideSensor_Unit3,MetalOxideSensor_Unit4,CO_GasSensor,Session ID,HVAC Operation Mode,Ambient Light Level,Activity Level
0,morning,19.63,53.83,125.486389,571.089440,478.554958,735.850412,654.625253,565.051969,2.0,7241,off,very_dim,Low Activity
1,morning,292.81,53.69,126.343018,575.789501,491.955151,740.282738,655.734327,557.078486,1.0,7241,ventilation_only,bright,Low Activity
2,morning,19.62,54.25,126.560695,585.543111,505.560808,737.112906,649.962421,558.065196,1.0,7241,eco_mode,None,Low Activity
3,morning,20.10,50.48,113.504877,597.449961,515.142272,744.020651,676.150075,600.222210,NaN,7241,eco_mode,very_bright,High Activity
4,morning,19.89,48.42,92.766225,613.654875,535.664558,770.265441,720.274019,625.731124,1.0,7241,heating_active,moderate,LowActivity


#TIME OF DAY

In [6]:
print(df['Time of Day'].value_counts())

Time of Day
afternoon    2586
evening      2432
morning      2415
night        2396
Name: count, dtype: int64


In [7]:
time_mapping = {
    'morning': 1,
    'afternoon': 2,
    'evening': 3,
    'night': 4
}

df['Time of Day'] = df['Time of Day'].map(time_mapping)

Assumed everything is okay in this column, data has been changed so that
1. 'morning': 1,
2. 'afternoon': 2,
3. 'evening': 3,
4. 'night': 4


#TEMPERATURE

In [8]:
print("Descriptive statistics for 'Temperature' column:")
print(df['Temperature'].describe())

print("\nMissing values in 'Temperature' column:")
print(df['Temperature'].isnull().sum())

Descriptive statistics for 'Temperature' column:
count    9829.000000
mean       42.618136
std        74.559993
min         6.903485
25%        18.930000
50%        20.510000
75%        23.180000
max       307.070000
Name: Temperature, dtype: float64

Missing values in 'Temperature' column:
0


Things taken note
1. Standard deviation is crazy, 74 degress celcius of houses is unreasonable.
2. max and min values is too unreasonable for housing
3. looking at the percentiles, majority of the data is between 18.9 to 23.1 degrees

Find acceptable range (IQR METHOD)
1. IQR is Q3 - Q1 (23.17 - 18.94 = 4.23) IQR = 4.23
2. Upper bound is 23.17 + (1.5 * 4.23) = 29.52
3. Lower bound is 18.94 - (1.5 * 4.23) = 12.59
4. acceptable range is 12.59 to 29.52

In [9]:
print(f"Original rows: {len(df)}")

lower_bound = 12.59
upper_bound = 29.52

df = df[(df['Temperature'] >= lower_bound) & (df['Temperature'] <= upper_bound)].copy()

print(f"Rows after filtering Temperature: {len(df)}")

Original rows: 9829
Rows after filtering Temperature: 8834


#Humidity

In [10]:
print("Descriptive statistics for 'Humidity' column:")
print(df['Humidity'].describe())

print("\nMissing values in 'Humidity' column:")
print(df['Humidity'].isnull().sum())

Descriptive statistics for 'Humidity' column:
count    7112.000000
mean       51.641706
std        21.154297
min       -49.870000
25%        47.710000
50%        51.410000
75%        53.840000
max       198.900000
Name: Humidity, dtype: float64

Missing values in 'Humidity' column:
1722


Humidity is only 0%-100%, anything else is too absurd, however, because this is data collected from homes, we are going to cut the humidity range to 20%-80%.

In [11]:
print(f"Original rows before Humidity filtering: {len(df)}")

lower_humidity_bound = 20
upper_humidity_bound = 80

df = df[(df['Humidity'] >= lower_humidity_bound) & (df['Humidity'] <= upper_humidity_bound)].copy()

print(f"Rows after filtering Humidity: {len(df)}")

Original rows before Humidity filtering: 8834
Rows after filtering Humidity: 6748


For the empty values will be settled later as we need to confirm the other column data is valid to move on to imputation

#CO2_InfraredSensor

In [12]:
print("Descriptive statistics for 'CO2_InfraredSensor' column:")
print(df['CO2_InfraredSensor'].describe())

print("\nMissing values in 'CO2_InfraredSensor' column:")
print(df['CO2_InfraredSensor'].isnull().sum())

Descriptive statistics for 'CO2_InfraredSensor' column:
count    6748.000000
mean      109.996635
std        34.086801
min       -17.472184
25%        96.980591
50%       112.966961
75%       124.634766
max       237.873938
Name: CO2_InfraredSensor, dtype: float64

Missing values in 'CO2_InfraredSensor' column:
0


In [13]:
from sklearn.preprocessing import MinMaxScaler

df['CO2_InfraredSensor'] = df['CO2_InfraredSensor'].clip(lower=0)
scaler = MinMaxScaler()

df[['CO2_InfraredSensor']] = scaler.fit_transform(df[['CO2_InfraredSensor']])

In [14]:
print("Descriptive statistics for 'CO2_InfraredSensor' column:")
print(df['CO2_InfraredSensor'].describe())

print("\nMissing values in 'CO2_InfraredSensor' column:")
print(df['CO2_InfraredSensor'].isnull().sum())

Descriptive statistics for 'CO2_InfraredSensor' column:
count    6748.000000
mean        0.462848
std         0.141821
min         0.000000
25%         0.407697
50%         0.474903
75%         0.523953
max         1.000000
Name: CO2_InfraredSensor, dtype: float64

Missing values in 'CO2_InfraredSensor' column:
0


We found impossible negative numbers in the CO2_InfraredSensor data, so we clipped them to zero to fix the hardware glitches.
The numbers do not match with regular ppm values, thus this data is messy and hard to understand, thus, we squished its values onto a 0-to-1 scale so the model only pays attention to sudden activity spikes instead of the inaccurate raw numbers

#CO2_ElectroChemicalSensor

In [15]:
print("Descriptive statistics for 'CO2_ElectroChemicalSensor' column:")
print(df['CO2_ElectroChemicalSensor'].describe())

print("\nMissing values in 'CO2_ElectroChemicalSensor' column:")
print(df['CO2_ElectroChemicalSensor'].isnull().sum())

Descriptive statistics for 'CO2_ElectroChemicalSensor' column:
count    6748.000000
mean      578.639743
std        22.519919
min       408.599386
25%       559.312212
50%       579.458325
75%       595.562365
max       637.895790
Name: CO2_ElectroChemicalSensor, dtype: float64

Missing values in 'CO2_ElectroChemicalSensor' column:
0


In [16]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()

df[['CO2_ElectroChemicalSensor']] = std_scaler.fit_transform(df[['CO2_ElectroChemicalSensor']])

In [17]:
print("Descriptive statistics for 'CO2_ElectroChemicalSensor' column:")
print(df['CO2_ElectroChemicalSensor'].describe())

print("\nMissing values in 'CO2_ElectroChemicalSensor' column:")
print(df['CO2_ElectroChemicalSensor'].isnull().sum())

Descriptive statistics for 'CO2_ElectroChemicalSensor' column:
count    6.748000e+03
mean     1.448884e-15
std      1.000074e+00
min     -7.551224e+00
25%     -8.583051e-01
50%      3.635195e-02
75%      7.515070e-01
max      2.631468e+00
Name: CO2_ElectroChemicalSensor, dtype: float64

Missing values in 'CO2_ElectroChemicalSensor' column:
0


The data is realistic here,(assuming this is ppm levels as it matches those numbers), no outlier clipping or imputation was required. However, to ensure parity across features for distance-based and linear models like Logistic Regression, a StandardScaler was applied to this column.

#MetalOxideSensor_Unit1

In [18]:
print("Descriptive statistics for 'MetalOxideSensor_Unit1' column:")
print(df['MetalOxideSensor_Unit1'].describe())

print("\nMissing values in 'MetalOxideSensor_Unit1' column:")
print(df['MetalOxideSensor_Unit1'].isnull().sum())

Descriptive statistics for 'MetalOxideSensor_Unit1' column:
count    6748.000000
mean      470.624800
std        50.961583
min       286.825662
25%       439.693074
50%       468.185610
75%       494.577980
max       632.891693
Name: MetalOxideSensor_Unit1, dtype: float64

Missing values in 'MetalOxideSensor_Unit1' column:
0


In [19]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
df[['MetalOxideSensor_Unit1']] = std_scaler.fit_transform(df[['MetalOxideSensor_Unit1']])

print("Descriptive statistics for 'MetalOxideSensor_Unit1' column after scaling:")
print(df['MetalOxideSensor_Unit1'].describe())

Descriptive statistics for 'MetalOxideSensor_Unit1' column after scaling:
count    6.748000e+03
mean     9.855780e-16
std      1.000074e+00
min     -3.606889e+00
25%     -6.070066e-01
50%     -4.786686e-02
75%      4.700591e-01
max      3.184338e+00
Name: MetalOxideSensor_Unit1, dtype: float64


#MetalOxideSensor_Unit2



In [20]:
print("Descriptive statistics for 'MetalOxideSensor_Unit2' column:")
print(df['MetalOxideSensor_Unit2'].describe())

print("\nMissing values in 'MetalOxideSensor_Unit2' column:")
print(df['MetalOxideSensor_Unit2'].isnull().sum())

Descriptive statistics for 'MetalOxideSensor_Unit2' column:
count    5782.000000
mean      728.288130
std        28.824198
min       611.168916
25%       709.738288
50%       726.873087
75%       744.837607
max       807.732943
Name: MetalOxideSensor_Unit2, dtype: float64

Missing values in 'MetalOxideSensor_Unit2' column:
966


In [21]:
print(f"Original rows before dropping NaNs in 'MetalOxideSensor_Unit2': {len(df)}")
df.dropna(subset=['MetalOxideSensor_Unit2'], inplace=True)
print(f"Rows after dropping NaNs in 'MetalOxideSensor_Unit2': {len(df)}")

Original rows before dropping NaNs in 'MetalOxideSensor_Unit2': 6748
Rows after dropping NaNs in 'MetalOxideSensor_Unit2': 5782


In [22]:
std_scaler = StandardScaler()
df[['MetalOxideSensor_Unit2']] = std_scaler.fit_transform(df[['MetalOxideSensor_Unit2']])

print("Descriptive statistics for 'MetalOxideSensor_Unit2' column after scaling:")
print(df['MetalOxideSensor_Unit2'].describe())

Descriptive statistics for 'MetalOxideSensor_Unit2' column after scaling:
count    5.782000e+03
mean    -1.572976e-15
std      1.000086e+00
min     -4.063577e+00
25%     -6.436067e-01
50%     -4.909640e-02
75%      5.742019e-01
max      2.756423e+00
Name: MetalOxideSensor_Unit2, dtype: float64


While imputation was considered, 970 rows were dropped to prevent inducing multicollinearity in the future models, prioritizing clean, independent spatial data over raw volume.

# MetalOxideSensor_Unit3

In [23]:
print("Descriptive statistics for 'MetalOxideSensor_Unit3' column:")
print(df['MetalOxideSensor_Unit3'].describe())

print("\nMissing values in 'MetalOxideSensor_Unit3' column:")
print(df['MetalOxideSensor_Unit3'].isnull().sum())

Descriptive statistics for 'MetalOxideSensor_Unit3' column:
count    5782.000000
mean      680.616603
std        55.128143
min       456.673900
25%       659.113280
50%       680.070425
75%       703.407600
max       905.801213
Name: MetalOxideSensor_Unit3, dtype: float64

Missing values in 'MetalOxideSensor_Unit3' column:
0


In [24]:
std_scaler = StandardScaler()
df[['MetalOxideSensor_Unit3']] = std_scaler.fit_transform(df[['MetalOxideSensor_Unit3']])

print("Descriptive statistics for 'MetalOxideSensor_Unit3' column after scaling:")
print(df['MetalOxideSensor_Unit3'].describe())

Descriptive statistics for 'MetalOxideSensor_Unit3' column after scaling:
count    5.782000e+03
mean    -2.428282e-15
std      1.000086e+00
min     -4.062572e+00
25%     -3.900944e-01
50%     -9.908277e-03
75%      4.134543e-01
max      4.085102e+00
Name: MetalOxideSensor_Unit3, dtype: float64


#MetalOxideSensor_Unit4

In [25]:
print("Descriptive statistics for 'MetalOxideSensor_Unit4' column:")
print(df['MetalOxideSensor_Unit4'].describe())

print("\nMissing values in 'MetalOxideSensor_Unit4' column:")
print(df['MetalOxideSensor_Unit4'].isnull().sum())

Descriptive statistics for 'MetalOxideSensor_Unit4' column:
count    5782.000000
mean      612.042790
std        42.753456
min       412.791904
25%       586.279643
50%       609.614584
75%       635.053958
max       733.131578
Name: MetalOxideSensor_Unit4, dtype: float64

Missing values in 'MetalOxideSensor_Unit4' column:
0


In [26]:
std_scaler = StandardScaler()
df[['MetalOxideSensor_Unit4']] = std_scaler.fit_transform(df[['MetalOxideSensor_Unit4']])

print("Descriptive statistics for 'MetalOxideSensor_Unit4' column after scaling:")
print(df['MetalOxideSensor_Unit4'].describe())

Descriptive statistics for 'MetalOxideSensor_Unit4' column after scaling:
count    5.782000e+03
mean     1.160070e-15
std      1.000086e+00
min     -4.660866e+00
25%     -6.026501e-01
50%     -5.680046e-02
75%      5.382760e-01
max      2.832502e+00
Name: MetalOxideSensor_Unit4, dtype: float64


#CO_GasSensor

In [27]:
print("Descriptive statistics for 'CO_GasSensor' column:")
print(df['CO_GasSensor'].describe())

print("\nMissing values in 'CO_GasSensor' column:")
print(df['CO_GasSensor'].isnull().sum())

Descriptive statistics for 'CO_GasSensor' column:
count    5290.000000
mean        1.296786
std         0.782688
min         0.000000
25%         1.000000
50%         1.000000
75%         2.000000
max         4.000000
Name: CO_GasSensor, dtype: float64

Missing values in 'CO_GasSensor' column:
492


In [28]:
print(f"Original rows before dropping NaNs in 'CO_GasSensor': {len(df)}")
df.dropna(subset=['CO_GasSensor'], inplace=True)
print(f"Rows after dropping NaNs in 'CO_GasSensor': {len(df)}")

from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
df[['CO_GasSensor']] = std_scaler.fit_transform(df[['CO_GasSensor']])

print("\nDescriptive statistics for 'CO_GasSensor' column after scaling:")
print(df['CO_GasSensor'].describe())

Original rows before dropping NaNs in 'CO_GasSensor': 5782
Rows after dropping NaNs in 'CO_GasSensor': 5290

Descriptive statistics for 'CO_GasSensor' column after scaling:
count    5.290000e+03
mean     7.521813e-17
std      1.000095e+00
min     -1.656993e+00
25%     -3.792244e-01
50%     -3.792244e-01
75%      8.985444e-01
max      3.454082e+00
Name: CO_GasSensor, dtype: float64


The data quite realistic, easy to use , just drop empty and standard scaler to normalise

#Session ID

In [29]:
print("Descriptive statistics for 'Session ID' column:")
print(df['Session ID'].describe())

print("\nMissing values in 'Session ID' column:")
print(df['Session ID'].isnull().sum())

Descriptive statistics for 'Session ID' column:
count    5290.000000
mean     5418.433270
std      2596.482285
min      1374.000000
25%      3074.000000
50%      5214.000000
75%      7395.000000
max      9658.000000
Name: Session ID, dtype: float64

Missing values in 'Session ID' column:
0


In [30]:
num_duplicates = df['Session ID'].duplicated().sum()
print(f"Number of duplicated Session IDs: {num_duplicates}")

Number of duplicated Session IDs: 5252


In [31]:
df.drop('Session ID', axis=1, inplace=True)
display(df.head())

,Time of Day,Temperature,Humidity,CO2_InfraredSensor,CO2_ElectroChemicalSensor,MetalOxideSensor_Unit1,MetalOxideSensor_Unit2,MetalOxideSensor_Unit3,MetalOxideSensor_Unit4,CO_GasSensor,HVAC Operation Mode,Ambient Light Level,Activity Level
0,1,19.63,53.83,0.527533,-0.335297,0.155622,0.262381,-0.471512,-1.099207,0.898544,off,very_dim,Low Activity
2,1,19.62,54.25,0.532049,0.306568,0.685587,0.306185,-0.556101,-1.262641,-0.379224,eco_mode,None,Low Activity
4,1,19.89,48.42,0.389981,1.554967,1.276345,1.456448,0.719430,0.320197,-0.379224,heating_active,moderate,LowActivity
5,2,19.82,47.71,0.334168,1.674640,1.713268,1.770494,1.056902,1.236118,-0.379224,Ventilation_Only,very_bright,Low Activity
6,4,20.04,48.42,0.497295,0.176963,0.081491,0.298479,0.259832,0.393972,-0.379224,heating_active,bright,Low Activity


Session ID has too many duplicates, cant drop the duplicates as too much data will be gone, and the id is quite useless for the model training

#HVAC Operation Mode

In [32]:
print("Value counts for 'HVAC Operation Mode' column:")
print(df['HVAC Operation Mode'].value_counts())

print("\nMissing values in 'HVAC Operation Mode' column:")
print(df['HVAC Operation Mode'].isnull().sum())

Value counts for 'HVAC Operation Mode' column:
HVAC Operation Mode
maintenance_mode    797
cooling_active      796
heating_active      785
off                 763
eco_mode            759
ventilation_only    702
COOLING_ACTIVE       63
VENTILATION_ONLY     59
HEATING_ACTIVE       57
Off                  53
MAINTENANCE_MODE     53
ECO_MODE             52
OFF                  47
Cooling_Active       39
Eco_Mode             35
Ventilation_Only     33
Heating_active       32
Eco_mode             31
Maintenance_mode     30
Cooling_active       28
Heating_Active       28
Ventilation_only     24
Maintenance_Mode     24
Name: count, dtype: int64

Missing values in 'HVAC Operation Mode' column:
0


In [33]:
df['HVAC Operation Mode'] = df['HVAC Operation Mode'].str.upper()
print("Uppercase 'HVAC Operation Mode' column values:")
print(df['HVAC Operation Mode'].value_counts())

Uppercase 'HVAC Operation Mode' column values:
HVAC Operation Mode
COOLING_ACTIVE      926
MAINTENANCE_MODE    904
HEATING_ACTIVE      902
ECO_MODE            877
OFF                 863
VENTILATION_ONLY    818
Name: count, dtype: int64


In [34]:
df = pd.get_dummies(df, columns=['HVAC Operation Mode'], prefix='HVAC_Mode', drop_first=True, dtype=int)
display(df.head())

,Time of Day,Temperature,Humidity,CO2_InfraredSensor,CO2_ElectroChemicalSensor,MetalOxideSensor_Unit1,MetalOxideSensor_Unit2,MetalOxideSensor_Unit3,MetalOxideSensor_Unit4,CO_GasSensor,Ambient Light Level,Activity Level,HVAC_Mode_ECO_MODE,HVAC_Mode_HEATING_ACTIVE,HVAC_Mode_MAINTENANCE_MODE,HVAC_Mode_OFF,HVAC_Mode_VENTILATION_ONLY
0,1,19.63,53.83,0.527533,-0.335297,0.155622,0.262381,-0.471512,-1.099207,0.898544,very_dim,Low Activity,0,0,0,1,0
2,1,19.62,54.25,0.532049,0.306568,0.685587,0.306185,-0.556101,-1.262641,-0.379224,None,Low Activity,1,0,0,0,0
4,1,19.89,48.42,0.389981,1.554967,1.276345,1.456448,0.719430,0.320197,-0.379224,moderate,LowActivity,0,1,0,0,0
5,2,19.82,47.71,0.334168,1.674640,1.713268,1.770494,1.056902,1.236118,-0.379224,very_bright,Low Activity,0,0,0,0,1
6,4,20.04,48.42,0.497295,0.176963,0.081491,0.298479,0.259832,0.393972,-0.379224,bright,Low Activity,0,1,0,0,0


To clean the HVAC Operation Mode data, we first converted all the text to uppercase so that the categories matched perfectly. Because machine learning models need numbers instead of words, we used One-Hot Encoding to turn each HVAC mode into its own column of 0s and 1s. Finally, we used the drop_first=True setting to remove one of the duplicate categories. This prevents a mathematical error called 'multicollinearity' (the dummy variable trap), which would otherwise confuse our Logistic Regression model."

#Ambient Light Level

In [35]:
print("Descriptive statistics for 'Ambient Light Level' column:")
print(df['Ambient Light Level'].describe())

print("\nMissing values in 'Ambient Light Level' column:")
print(df['Ambient Light Level'].isnull().sum())

print("\nValue counts for 'Ambient Light Level' column:")
print(df['Ambient Light Level'].value_counts())

Descriptive statistics for 'Ambient Light Level' column:
count            4698
unique              5
top       very_bright
freq             1773
Name: Ambient Light Level, dtype: object

Missing values in 'Ambient Light Level' column:
592

Value counts for 'Ambient Light Level' column:
Ambient Light Level
very_bright    1773
bright         1399
moderate        762
dim             470
very_dim        294
Name: count, dtype: int64


In [36]:
print(f"Original rows before dropping NaNs in 'Ambient Light Level': {len(df)}")
df.dropna(subset=['Ambient Light Level'], inplace=True)
print(f"Rows after dropping NaNs in 'Ambient Light Level': {len(df)}")
display(df.head())

Original rows before dropping NaNs in 'Ambient Light Level': 5290
Rows after dropping NaNs in 'Ambient Light Level': 4698


,Time of Day,Temperature,Humidity,CO2_InfraredSensor,CO2_ElectroChemicalSensor,MetalOxideSensor_Unit1,MetalOxideSensor_Unit2,MetalOxideSensor_Unit3,MetalOxideSensor_Unit4,CO_GasSensor,Ambient Light Level,Activity Level,HVAC_Mode_ECO_MODE,HVAC_Mode_HEATING_ACTIVE,HVAC_Mode_MAINTENANCE_MODE,HVAC_Mode_OFF,HVAC_Mode_VENTILATION_ONLY
0,1,19.63,53.83,0.527533,-0.335297,0.155622,0.262381,-0.471512,-1.099207,0.898544,very_dim,Low Activity,0,0,0,1,0
4,1,19.89,48.42,0.389981,1.554967,1.276345,1.456448,0.719430,0.320197,-0.379224,moderate,LowActivity,0,1,0,0,0
5,2,19.82,47.71,0.334168,1.674640,1.713268,1.770494,1.056902,1.236118,-0.379224,very_bright,Low Activity,0,0,0,0,1
6,4,20.04,48.42,0.497295,0.176963,0.081491,0.298479,0.259832,0.393972,-0.379224,bright,Low Activity,0,1,0,0,0
7,4,19.29,52.67,0.572962,-0.200646,0.084385,-0.239676,-0.001916,0.239353,-0.379224,bright,Moderate Activity,0,0,1,0,0


In [37]:
light_mapping = {
    'very_dim': 0,
    'dim': 1,
    'moderate': 2,
    'bright': 3,
    'very_bright': 4
}

df['Ambient Light Level'] = df['Ambient Light Level'].map(light_mapping)

The Ambient Light Level feature presented as ordinal categorical data with 594 missing values. We just dropped them. Because the categories possess a definitive inherent ranking, Ordinal Encoding was utilized rather than One-Hot Encoding. The categories were mapped to numerical integers (0 to 4), preserving the linear relationship of light intensity for the baseline Logistic Regression model.

#Activity Level

In [38]:
print("Descriptive statistics for 'Activity Level' column:")
print(df['Activity Level'].describe())

print("\nMissing values in 'Activity Level' column:")
print(df['Activity Level'].isnull().sum())

print("\nValue counts for 'Activity Level' column:")
print(df['Activity Level'].value_counts())

Descriptive statistics for 'Activity Level' column:
count             4698
unique               6
top       Low Activity
freq              2438
Name: Activity Level, dtype: object

Missing values in 'Activity Level' column:
0

Value counts for 'Activity Level' column:
Activity Level
Low Activity         2438
Moderate Activity    1310
High Activity         523
ModerateActivity      162
Low_Activity          140
LowActivity           125
Name: count, dtype: int64


In [39]:
target_mapping = {
    'Low Activity': 0,
    'Low_Activity': 0,
    'LowActivity': 0,
    'Moderate Activity': 1,
    'ModerateActivity': 1,
    'High Activity': 2
}


df['Activity Level'] = df['Activity Level'].map(target_mapping)

Analysis of the target variable, Activity Level, revealed data entry inconsistencies resulting in duplicated categorical classes (e.g., 'Low Activity', 'Low_Activity'). A mapping dictionary was utilized to consolidate these typographic errors and simultaneously apply Ordinal Encoding, transforming the target into a numerical scale (0 for Low, 1 for Moderate, 2 for High). Post-consolidation, a class imbalance was identified, with 'High Activity' representing only 11% of the dataset. This highlights the necessity of evaluating subsequent predictive models using Recall and F1-scores rather than baseline accuracy, and integrating class-weight balancing during model training

#Final Analysis

based on what we see, the data is quite skewed towards people with lower activity, this is something to take note for later during model making.

Ambient light level is also skewed towards brighter

hvac is quite even, but dropped the cooling active for the prevention of perfect multicollinearity

session id is useless

temp and humidity is done to assume regular values

word values are either one hot encoded or label encoding